### PR TITLE
Add a setjmp/longjmp runtime for a new sjlj translation proposed in LLVM

### PR DESCRIPTION
**Pull request by yamt**
the new runtime code is basically a copy from:
https://github.com/yamt/garbage/blob/wasm-sjlj-alt2/wasm/longjmp/rt.c.

The corresponding LLVM change:
https://github.com/llvm/llvm-project/pull/84137

Discussion:
https://docs.google.com/document/d/1ZvTPT36K5jjiedF8MCXbEmYjULJjI723aOAks1IdLLg/edit

### CODE REVIEW

**Comment by yamt** [comment]
to use this, i suppose we need one of the following:

* make the LLVM use the new method unconditionally
* or, make some changes in SUPPORT_LONGJMP and tools/building.py to specify the new LLVM option

**Review by sbc100 — COMMENTED** [review]
Thanks for working on this!

Would you mind adding these new functions to the existing `emscripten_setjmp.c` since that is where these functions currently live?

Also, can you match the formatting rules of that file which is basically LLVM coding style (you can use clang-format to do this for you I think).

File: system/lib/compiler-rt/wasm_setjmp.c
Comment (sbc100) [inline system/lib/compiler-rt/wasm_setjmp.c:44]: Maybe replace these two with you replace these with `assert(...)` ?

File: system/lib/compiler-rt/wasm_setjmp.c
Comment (sbc100) [inline system/lib/compiler-rt/wasm_setjmp.c:43]: Should we just declare the first argument as `struct jmp_buf_impl *jb`? (i.e. why case from void* here?)

File: system/lib/compiler-rt/wasm_setjmp.c
Comment (sbc100) [inline system/lib/compiler-rt/wasm_setjmp.c:35]: Why not declare this as `struct jmp_buf_impl*` ?

**Review by aheejin — COMMENTED** [review]
Thanks! I have basically the same set of comments with @sbc100:
1. clang-format with our current style
2. We seem to use `//` style comments in our existing libraries. So maybe match that for consistency
3. I guess it'd be better to move it to the current file `emscripten-setjmp.c` at least for now. We can discuss more on where to move this for the upstreaming later, but I don't think we need to be blocked on that.

Also, unless you are planning for more changes, we can un-draft this PR I think, given that we want to land this before the LLVM PR?

File: system/lib/compiler-rt/wasm_setjmp.c
Comment (aheejin) [inline system/lib/compiler-rt/wasm_setjmp.c:12]: I don't think this comment is necessary. Also, `__wasm_sjlj_setjmp` and `__wasm_sjlj_test` can be used with Emscripten SjLj too. Only `__wasm_sjlj_longjmp`, which uses Wasm `throw`, is exclusive to Wasm SjLj.

File: system/lib/compiler-rt/wasm_setjmp.c
Comment (aheejin) [inline system/lib/compiler-rt/wasm_setjmp.c:22]: We seem to use `//` with comments.
Nit: Can you wrap to 80 cols? (This and other comment lines too) This looks a bit too short.

File: system/lib/compiler-rt/wasm_setjmp.c
Comment (aheejin) [inline system/lib/compiler-rt/wasm_setjmp.c:15]: As I said above `setjmp` and `test` can be shared with Emscripten SjLj, so we shouldn't have this.

File: system/lib/compiler-rt/wasm_setjmp.c
Comment (aheejin) [inline system/lib/compiler-rt/wasm_setjmp.c:78]: Do we need these `>`s?

File: system/lib/compiler-rt/wasm_setjmp.c
Comment (aheejin) [inline system/lib/compiler-rt/wasm_setjmp.c:9]: I don't think we need to have these links in the library

**Review by aheejin — COMMENTED** [review]
+ Some nits.

Also, I'd like to bikeshed a bit on the (final) names of the functions... I agree it's good to prefix all functions with `__wasm_`, especially for future upstreaming. Do we need `_sjlj_` too, given that `setjmp` and `longjmp` already imply that?

I like these, WDYT?
- `__wasm_setjmp`
- `__wasm_test_setjmp`
- `__wasm_longjmp`
  - For now this can't be done because we have to have the current `__wasm_longjmp`. I suggest we land with a different name (the current `__wasm_sjlj_longjmp` seems fine) and change this (both in LLVM and emscripten) later.

**Comment by sbc100** [comment]
> Some nits.
> 
> Also, I'd like to bikeshed a bit on the (final) names of the functions... I agree it's good to prefix all functions with `__wasm_`, especially for future upstreaming. Do we need `_sjlj_` too, given that `setjmp` and `longjmp` already imply that?
> 
> I like these, WDYT?
> 
> * `__wasm_setjmp`
> * `__wasm_test_setjmp`

How about `__wasm_setjmp_test` for this one?

**Comment by aheejin** [comment]
`__wasm_setjmp_test` sounds good too.

File: system/lib/compiler-rt/wasm_setjmp.c
Comment (yamt) [reply system/lib/compiler-rt/wasm_setjmp.c:43]: i don't want to have jmp_buf_impl in the function prototype as it's an internal detail.

**Comment by yamt** [comment]
>  For now this can't be done because we have to have the current __wasm_longjmp. I suggest we land with a different name (the current __wasm_sjlj_longjmp seems fine) and change this (both in LLVM and emscripten) later.

i just used the same name as the change in this function is not essential.

File: system/lib/compiler-rt/emscripten_setjmp.c
Comment (sbc100) [inline system/lib/compiler-rt/emscripten_setjmp.c:129]: Does the corresponding LLVM change not also change the ABI for emscripten exceptions too?  i.e. not just when __USING_WASM_SJLJ__?  (i.e. should all this new code live outside the `#if` block?

File: system/lib/compiler-rt/wasm_setjmp.c
Comment (sbc100) [reply system/lib/compiler-rt/wasm_setjmp.c]: Fair enough.   I was thinking that because this function is not actually declared anywhere else then it would not matter since this internal detail would remains solely within this `.c` file.

File: system/lib/compiler-rt/emscripten_setjmp.c
Comment (aheejin) [reply system/lib/compiler-rt/emscripten_setjmp.c:129]: These `__wasm_setjmp` and `__wasm_setjmp_test` can be used both for emscripten and Wasm SjLj. Actually, using different functions for the two of them will be messy and not allow us to delete the existing `saveSetjmp` and `testSetjmp`. This is currently within `#ifdef __USING_WASM_SJLJ__`. Only `__wasm_longjmp` is exclusive to Wasm SjLj.

**Comment by aheejin** [comment]
> > For now this can't be done because we have to have the current __wasm_longjmp. I suggest we land with a different name (the current __wasm_sjlj_longjmp seems fine) and change this (both in LLVM and emscripten) later.
> 
> i just used the same name as the change in this function is not essential.

Does this mean you deleted the `__wasm_longjmp`? I can't find it in the diff anymore. Then do you have a plan to re-add it? If they are functionally the same, I think you can just replace it.

File: system/lib/compiler-rt/emscripten_setjmp.c
Comment (yamt) [reply system/lib/compiler-rt/emscripten_setjmp.c:129]: > Does the corresponding LLVM change not also change the ABI for emscripten exceptions too? i.e. not just when **USING_WASM_SJLJ**? (i.e. should all this new code live outside the `#if` block?

currently, no.

Comment (yamt) [reply system/lib/compiler-rt/emscripten_setjmp.c:129]: > +1. These `__wasm_setjmp` and `__wasm_setjmp_test` can be used both for emscripten and Wasm SjLj. Actually, using different functions for the two of them will be messy and not allow us to delete the existing `saveSetjmp` and `testSetjmp`. This is currently within `#ifdef __USING_WASM_SJLJ__`. Only `__wasm_longjmp` is exclusive to Wasm SjLj.

maybe. i haven't looked at the emscrpiten sjlj closely. do you think it's enough to apply the same change mechanically?

**Comment by yamt** [comment]
> > > For now this can't be done because we have to have the current __wasm_longjmp. I suggest we land with a different name (the current __wasm_sjlj_longjmp seems fine) and change this (both in LLVM and emscripten) later.
> > 
> > 
> > i just used the same name as the change in this function is not essential.
> 
> Does this mean you deleted the `__wasm_longjmp`? I can't find it in the diff anymore.

yes.

> Then do you have a plan to re-add it? If they are functionally the same, I think you can just replace it.

ok.

File: system/lib/compiler-rt/emscripten_setjmp.c
Comment (sbc100) [inline system/lib/compiler-rt/emscripten_setjmp.c:125]: Is this code going to work both before and after the llvm change?

Comment (yamt) [reply system/lib/compiler-rt/emscripten_setjmp.c:125]: > Is this code going to work both before and after the llvm change?

i expect so.
at least it passed ./test/runner.py 'other.test_setjmp*' and ./test/runner.py 'other.test_setjmp*'.
do you have specific concerns?

**Comment by yamt** [comment]
@sbc100 can you advise me about the latest CI failure? it isn't obvious to me what they are complaining about.
```
======================================================================
FAIL [0.005s]: test_minimal_runtime_code_size_hello_webgl2_wasm2js (test_other.other)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/distiller/emsdk/python/3.9.2_64bit/lib/python3.9/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/Users/distiller/emsdk/python/3.9.2_64bit/lib/python3.9/unittest/case.py", line 593, in run
    self._callTestMethod(testMethod)
  File "/Users/distiller/emsdk/python/3.9.2_64bit/lib/python3.9/unittest/case.py", line 550, in _callTestMethod
    method()
  File "/Users/distiller/project/test/common.py", line 630, in resulting_test
    return func(self, *args)
  File "/Users/distiller/project/test/test_other.py", line 10841, in test_minimal_runtime_code_size
    self.assertEqual(total_output_size, total_expected_size)
  File "/Users/distiller/emsdk/python/3.9.2_64bit/lib/python3.9/unittest/case.py", line 831, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/Users/distiller/emsdk/python/3.9.2_64bit/lib/python3.9/unittest/case.py", line 824, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 22635 != 22627

======================================================================
FAIL [0.004s]: test_minimal_runtime_code_size_hello_webgl_wasm2js (test_other.other)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/distiller/emsdk/python/3.9.2_64bit/lib/python3.9/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/Users/distiller/emsdk/python/3.9.2_64bit/lib/python3.9/unittest/case.py", line 593, in run
    self._callTestMethod(testMethod)
  File "/Users/distiller/emsdk/python/3.9.2_64bit/lib/python3.9/unittest/case.py", line 550, in _callTestMethod
    method()
  File "/Users/distiller/project/test/common.py", line 630, in resulting_test
    return func(self, *args)
  File "/Users/distiller/project/test/test_other.py", line 10841, in test_minimal_runtime_code_size
    self.assertEqual(total_output_size, total_expected_size)
  File "/Users/distiller/emsdk/python/3.9.2_64bit/lib/python3.9/unittest/case.py", line 831, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/Users/distiller/emsdk/python/3.9.2_64bit/lib/python3.9/unittest/case.py", line 824, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 22137 != 22129
```

**Comment by sbc100** [comment]
Those "code_size" test failures are likely unrelated.  They often happen then llvm changes effect code size.   Can you trying rebasing or merging?

File: system/lib/compiler-rt/emscripten_setjmp.c
Comment (sbc100) [reply system/lib/compiler-rt/emscripten_setjmp.c:125]: I guess I don't see how this same function can work both before and after the llvm change.   Won't llvm generate different `env` arguments before and after your change?   Is the layout of `env` not changing with your llvm change?

Can you also run `./test/runner core0.test_setjmp*`?

Comment (yamt) [reply system/lib/compiler-rt/emscripten_setjmp.c:125]: it passed `./test/runner core0.test_setjmp*` as well. (it was what i meant. repeating two `other.test_setjmp*` was a copy-and-paste mistake.)

the original jmp_buf was large enough to store &jb->arg. it was just unused before this change.

**Comment by yamt** [comment]
> Those "code_size" test failures are likely unrelated. They often happen then llvm changes effect code size. Can you trying rebasing or merging?

ok. let me try a rebase. i hope the force push won't make @aheejin too unhappy.

**Review by sbc100 — APPROVED** [review]
Awesome!  I'm excited to see this land

**Comment by aheejin** [comment]
> > Those "code_size" test failures are likely unrelated. They often happen then llvm changes effect code size. Can you trying rebasing or merging?
> 
> ok. let me try a rebase. i hope the force push won't make @aheejin too unhappy.

You can do a merge (instead of a rebase) via `git merge main` in order not to force-push 😀 But anyway, I think it's fine either way in small PRs like this. Sorry if that caused any inconvenience.

**Review by sbc100 — COMMENTED** [review]
This lgtm now.   @aheejin, any more comments?

File: system/lib/compiler-rt/emscripten_setjmp.c
Comment (aheejin) [inline system/lib/compiler-rt/emscripten_setjmp.c:86]: Why delete this?

File: system/lib/compiler-rt/emscripten_setjmp.c
Comment (aheejin) [inline system/lib/compiler-rt/emscripten_setjmp.c:125]: Why would we delete `C_LONGJMP` define and replace it with 1?

File: system/lib/compiler-rt/emscripten_setjmp.c
Comment (aheejin) [inline system/lib/compiler-rt/emscripten_setjmp.c:119]: I know it's preexisting comment, but how about changing it to `//` style comment so that it's consistent with other comments?

File: system/lib/compiler-rt/emscripten_setjmp.c
Comment (aheejin) [inline system/lib/compiler-rt/emscripten_setjmp.c:90]: Why delete this comment?

File: system/lib/compiler-rt/emscripten_setjmp.c
Comment (aheejin) [inline system/lib/compiler-rt/emscripten_setjmp.c:82]: That we don't need this `thread_local struct` is nice!

File: system/lib/compiler-rt/emscripten_setjmp.c
Comment (yamt) [reply system/lib/compiler-rt/emscripten_setjmp.c]: i replaced the __wasm_longjmp implementation with mine as you suggested and this and other comment changes were a part of it.
anyway, restored.

**Review by aheejin — COMMENTED** [review]
I ran all sjlj tests with this PR and https://github.com/llvm/llvm-project/pull/84137 and all tests seem to pass except for [`test_longjmp_zero`](https://github.com/emscripten-core/emscripten/blob/918e131fae0b5c7b1d05a5c75d7e8e676c377713/test/test_core.py#L895-L900) for Emscripten SjLj. Turns out `emscripten_longjmp` needs to adjust `val` to 1 when the input is 0 too. LGTM otherwise.

---

Then the question is why `test_longjmp_zero` was passing for emscripten SjLj until now. I investigated it a little, and the answer is, accidentally.

Feel free to skip the below if not interested. I'm writing this mostly for my own record-keeping.

https://github.com/llvm/llvm-project/blob/99a8ffe6b1e78d16f5dc528048a7c34a236b62bb/llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp#L157-L169
In this code, it checks whether both `__THREW__` and `__threwValue` are nonzero. Nonzero `__THREW__` means an exception or a longjmp occurred. `__threwValue` is the second argument to `longjmp`. Somehow it checks whether `__threwValue` is 0, and if so, goes here:
https://github.com/llvm/llvm-project/blob/99a8ffe6b1e78d16f5dc528048a7c34a236b62bb/llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp#L166-L167
Which causes the `switch` to go to the splitted next BB, which means, run as if a longjmp hasn't occurred:
https://github.com/llvm/llvm-project/blob/99a8ffe6b1e78d16f5dc528048a7c34a236b62bb/llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp#L174

But I don't think this is correct. `longjmp` has occurred if `__THREW__` is set, and is not dependent on whether `__threwValue` is 0 or not, but this code assumes `longjmp` occurred only when both of them are nonzero. I don't know why I wrote code like this, but I guess I was porting then fastcomp's implementation and moved this part just verbatim: https://github.com/emscripten-core/emscripten-fastcomp/blob/9bdc7bb4fc595fe05a021b06fe350e8494a741a1/lib/Target/JSBackend/CallHandlers.h#L274-L278

And if you run the `else` part, you don't end up running this `setTempRet0`: https://github.com/llvm/llvm-project/blob/99a8ffe6b1e78d16f5dc528048a7c34a236b62bb/llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp#L165

And the result of longjmp is read here, but because we haven't run `setTempRet0`, the value we read will be whatever value that was set before:
https://github.com/llvm/llvm-project/blob/99a8ffe6b1e78d16f5dc528048a7c34a236b62bb/llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp#L169

And before, we happened to call `setTempRet0` with the `setjmpTableSize`. Not sure why we were doing that. Even the comment says "unneeded": https://github.com/emscripten-core/emscripten/blob/918e131fae0b5c7b1d05a5c75d7e8e676c377713/system/lib/compiler-rt/emscripten_setjmp.c#L48

So as long as we call `longjmp(buf, 0)` after `setjmp`, if the second argument to `longjmp` is 0, we end up getting 4, which is `setjmpTableSize`, as a result. But this PR and the companion llvm PR removes `setjmpTable` and `setjmpTableSize` and the `setTempRet0` call with them. So now `getTempRet0` here ends up getting 0 (not because it was `longjmp`'s second argument but because it was read before being set with any values)
https://github.com/llvm/llvm-project/blob/99a8ffe6b1e78d16f5dc528048a7c34a236b62bb/llvm/lib/Target/WebAssembly/WebAssemblyLowerEmscriptenEHSjLj.cpp#L169

So `test_longjmp_zero.c`'s this line returns 0: https://github.com/emscripten-core/emscripten/blob/918e131fae0b5c7b1d05a5c75d7e8e676c377713/test/core/test_longjmp_zero.c#L14
and the next line `if (val)` is not taken, and goes to `longjmp` again, and `setjmp` return 0 again, ... so it goes into an infinite loop and hangs.

**Comment by sbc100** [comment]
Thanks for the details analysis of the `test_longjmp_zero` issue @aheejin.

When I added that test I did notice the setjmp was returning 4 but I couldn't figure out why.

Do you think we could fix that issue first, before we land the ABI change?   I assume it would be an standalone llvm change, at which point we could update the `test_longjmp_zero` test to actually assert that the return value is 1.

**Comment by aheejin** [comment]
No I think we should land this first, or add 
```c
if (val ==0)
  val = 1
```
separately to `emscripten_longjmp` first. What we do in LLVM is to add an unnecessary condition `& __threwValue = 0)` to the `longjmp` condition, but with the correct library this can't happen anyway. I'll fix LLVM anyway after these SjLj PRs land because it is unnecessary code.

**Comment by sbc100** [comment]
> No I think we should land this first, or add
> 
> ```c
> if (val == 0)
>   val = 1
> ```
> 
> separately to `emscripten_longjmp` first. 

That sounds good.  I couldn't find a place to do this myself when I added `test_longjmp_zero` .. it would be good get that done and add the extra assert to `test_longjmp_zero`.

**Comment by yamt** [comment]
> > No I think we should land this first, or add
> > ```c
> > if (val == 0)
> >   val = 1
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > separately to `emscripten_longjmp` first.
> 
> That sounds good. I couldn't find a place to do this myself when I added `test_longjmp_zero` .. it would be good get that done and add the extra assert to `test_longjmp_zero`.

https://github.com/emscripten-core/emscripten/pull/21577

**Review by aheejin — APPROVED** [review]
So I think we can merge this now? Will merge it. Thank you for working on this!

**Comment by aheejin** [comment]
He pushed another conflicts-resolving commit so the LLVM CI is still running, so I think we can do that some hours later.

**Pull request merged**
